app/Http/Controllers/PilotPassportController.php

```diff
@@ -0,0 +1,193 @@
+            $e->cid = $p->id;
+            $e->challenge_id = $c->id;
+            $e->save();
+            Mail::to($p->email)->send(new PilotPassportMail('enroll', $p, $c));
+            return redirect(route('pilotPassportIndex'))->withInput(['tab' => 'enrollments'])->with('success', 'You are now enrolled in the ZTL Pilot Passport program!');
+        }
+        return redirect(route('pilotPassportIndex'))->withInput(['tab' => 'enrollments'])->with('error', 'You are already enrolled in this challenge.');
+    }
+
+    public function setPrivacy(Request $request) {
+        $p = auth()->guard('realops')->user();
+        if (is_null($p)) {
+            return redirect(route('pilotPassportIndex'))->withInput(['tab' => 'settings'])->with('error', 'Unable to adjust privacy settings - Invalid CID.');
+        }
+        $p->privacy = $request->privacy;
+        $p->save();
+        return redirect(route('pilotPassportIndex'))->withInput(['tab' => 'settings'])->with('success', 'Your privacy preferences have been saved.');
+    }
+
+    public function purgeData(Request $request) {
+        if (strtolower($request->input('confirm_text')) != "confirm - purge all") {
+            return redirect()->back()->with('error', 'Data not purged. Please type in the required message to continue');
+        }
+        $p = auth()->guard('realops')->user();
+        $p->delete();
+        $enrollments = PilotPassportEnrollment::where('cid', $request->id);
+        foreach($enrollments as $e) {
+            $e->delete();
+        }
+        $logs = PilotPassportLog::where('cid', $request->id);
+        foreach($logs as $l) {
+            $l->delete();
+        }
+        return redirect('/logout');
+    }
+
+    public static function fetchRecentPilotAccomplishments() {
+        $accomplishments = PilotPassportAward::where('awarded_on', '>', now()->subDays(90)->endOfDay())
+            ->orderByRaw('RAND()')->take(10)->get();
+        $ret = null;
+        foreach ($accomplishments as $acc) {
+            $a = (object) [
+                'pilot_name' => $acc->pilot_public_name,
+                'challenge_name' => $acc->challenge_title
+            ];
+            $ret[] = $a;
+        }
+        return $ret;
+    }
+
+    public function generateCertificate($id) {
+        $a = PilotPassportAward::find($id);
+        $error_html = '<p>An error has occured - please contact <a href="emailto:wm@ztlartcc.org">wm@ztlartcc.org</a></p>';
+        if (!$a) {
+            return pdf()->html($error_html)->download();
+        }
+        $p = RealopsPilot::find($a->cid);
+        $staff_users = User::with('roles')->where('status', '1')->get();
+        $atm = $staff_users->filter(function ($staff_user) {
+            return $staff_user->hasRole('atm');
+        });
+        if (!$atm) {
+            $atm = $staff_users->filter(function ($staff_user) {
+                return $staff_user->hasRole('datm');
+            });
+        }
+        $atm = $atm->first();
```

Comment:
We should be able to do this when pulling from the DB to prevent having to do a query and then filter ourselves (and filtering a query is generally faster)

https://laratrust.santigarcor.me/docs/6.x/usage/querying-relationships.html#by-role

Reply:
Ahh okay that's frustrating. For a future story, how difficult would it be to restructure this? Is it like an update or is it an entire delete the User model and start from scratch?